tmt/steps/provision/mock.py

```diff
@@ -0,0 +1,505 @@
+                # shell is ready
+                break
+            for fileno, _ in events:
+                if fileno == self.mock_shell_stderr_fd:
+                    for line in self.mock_shell.stderr.readlines():
+                        self.parent._logger.debug("err", line.rstrip(), 'yellow', level = 0)
+
+        if mock_shell_result is not None:
+            raise tmt.utils.ProvisionError("Failed to launch mock shell: exited {}".format(mock_shell_result))
+
+        self.parent.verbose("Mock shell is ready")
+
+        # We do not expect these commands to fail.
+        self.mock_shell.stdin.write("rm -rf /tmp/stdout /tmp/stderr /tmp/returncode\n")
+        self.mock_shell.stdin.write("mkfifo /tmp/stdout /tmp/stderr /tmp/returncode\n")
+        self.mock_shell.stdin.flush()
+
+        loop = 2
+        while loop != 0 and self.mock_shell.poll() is None:
+            events = self.epoll.poll()
+            for fileno, _ in events:
+                if fileno == self.mock_shell_stdout_fd:
+                    loop -= 1
+                    self.mock_shell.stdout.read()
+                    break
+        self.mock_shell.stderr.read()
+
+        """
+        NOTE Here we would like to add code which unmounts the overlayfs
+        after the prepare phase is done. Nevertheless testing should work even
+        with the bootstrap chroot mounted.
+
+        plan = self.parent.parent.parent
+
+        prepare_data_class = cast(  # type: ignore[redundant-cast]
+            type[tmt.steps.prepare.shell.PrepareShellData],
+            tmt.steps.prepare.shell.PrepareShell.get_data_class(),
+        )
+
+        data = prepare_data_class(
+            name="tmt-unmount-mock-bootstrap-overlay",
+            how='shell',
+            script=[
+                TODO
+            ],
+        )
+
+        phase: PreparePlugin[Any] = cast(
+            PreparePlugin[Any],
+            PreparePlugin.delegate(plan.prepare, data=data),
+        )
+
+        plan.prepare._phases.append(phase)
+        """
+
+        return self
+
+    def execute(
+        self,
+        command: Command,
+        cwd: Optional[Path] = None,
+        env: Optional[tmt.utils.Environment] = None,
+        friendly_command: str = None,
+        log: Optional[tmt.log.LoggingFunction] = None,
+        silent: bool = False,
+    ):
+        """
+        Execute the command in a running mock shell for increased speed.
+        """
+
+        if self.mock_shell is None:
+            self.enter_shell()
+        shell_command = ""
+        if cwd is not None:
+            shell_command += "(cd " + shlex.quote(str(cwd)) + " && "
+        if env is not None:
+            for key, value in env.items():
+                shell_command += f"{key}={shlex.quote(value)} "
+        shell_command += str(command)
+        if cwd is not None:
+            shell_command += ")"
+        shell_command += " 1>/tmp/stdout 2>/tmp/stderr; echo $?>/tmp/returncode"
```

Comment:
Too many leading and trailing spaces for my taste. And round brackets. Easy to forget one here and there. How about moving the spaces and round bracket handling to one place? Something like this.

```python
shell_command_components: list[str] = [
    command
]

if env is not None:
    shell_command_components = [
        *(f"{key}={shlex.quote(value)}" for key, value in env.items()),
        *shell_command_components
    ]

if cwd is not None:
    shell_command_components = [
        "(",
        f"cd {shlex.quote(str(cwd))}",
        "&&",
        *shell_command_components,
        ")"
    ]

shell_command_components = [
    *shell_command_components,
    "1>/tmp/stdout 2>/tmp/stderr; echo $?"
]

shell_command = ' '.join(shell_command_components)
```

Reply:
Yes, looks good.